File: products/LoanHub/apps/backend/database/schemas/performance.py
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class EmployeePerformanceRead(BaseModel):
    staff_id: UUID
    employee_id: UUID | None = None
    user_id: UUID
    branch_id: UUID | None = None
    employee_name: str
    role: str
    job_title: str | None = None
    department: str | None = None
    offers_created: int
    offers_accepted: int
    loans_approved: int
    loans_disbursed: int
    payment_transactions: int
    successful_payment_amount: Decimal
    active_goals: int
    completed_goals: int
    goal_completion_percent: float
    average_review_score: float
    performance_score: float


class BranchPerformanceRead(BaseModel):
    branch_id: UUID | None = None
    branch_name: str
    employee_count: int
    active_loans: int
    overdue_loans: int
    loan_principal: Decimal
    payments_received: Decimal
    average_employee_score: float


class CompanyPerformanceRead(BaseModel):
    company_id: UUID
    company_name: str
    employee_count: int
    branch_count: int
    active_loans: int
    overdue_loans: int
    outstanding_balance: Decimal
    successful_payments: Decimal
    employee_average_score: float
    operational_score: float


class PerformanceOverviewRead(BaseModel):
    scope: str
    employee_count: int
    branch_count: int
    company_count: int
    active_loans: int
    overdue_loans: int
    total_outstanding: Decimal
    successful_payments: Decimal
    average_employee_score: float
    employees: list[EmployeePerformanceRead] = Field(default_factory=list)
    branches: list[BranchPerformanceRead] = Field(default_factory=list)
    companies: list[CompanyPerformanceRead] = Field(default_factory=list)

File: products/LoanHub/apps/frontend/types/performance.ts
export type EmployeePerformance = {
    staff_id: string;
    employee_id: string | null;
    user_id: string;
    branch_id: string | null;
    employee_name: string;
    role: string;
    job_title: string | null;
    department: string | null;
    offers_created: number;
    offers_accepted: number;
    loans_approved: number;
    loans_disbursed: number;
    payment_transactions: number;
    successful_payment_amount: string | number;
    active_goals: number;
    completed_goals: number;
    goal_completion_percent: number;
    average_review_score: number;
    performance_score: number;
};

export type BranchPerformance = {
    branch_id: string | null;
    branch_name: string;
    employee_count: number;
    active_loans: number;
    overdue_loans: number;
    loan_principal: string | number;
    payments_received: string | number;
    average_employee_score: number;
};

export type CompanyPerformance = {
    company_id: string;
    company_name: string;
    employee_count: number;
    branch_count: number;
    active_loans: number;
    overdue_loans: number;
    outstanding_balance: string | number;
    successful_payments: string | number;
    employee_average_score: number;
    operational_score: number;
};

export type PerformanceOverview = {
    scope: "company" | "platform" | string;
    employee_count: number;
    branch_count: number;
    company_count: number;
    active_loans: number;
    overdue_loans: number;
    total_outstanding: string | number;
    successful_payments: string | number;
    average_employee_score: number;
    employees: EmployeePerformance[];
    branches: BranchPerformance[];
    companies: CompanyPerformance[];
};
